libs/api/ebpf_api.cpp

```diff
@@ -4371,136 +4371,178 @@ ebpf_program_bind_map(fd_t program_fd, fd_t map_fd) NO_EXCEPT_TRY
+    }
+
+    std::mutex lock;
+    _Write_guarded_by_(lock) boolean unsubscribed;
+    ebpf_handle_t map_handle;
+    void* callback_context;
+    ring_buffer_sample_fn ring_buffer_sample_callback;
+    perf_buffer_lost_fn lost_callback;
+    perf_buffer_sample_fn perf_buffer_sample_callback;
+    uint32_t key_size;
+    uint32_t value_size;
+    uint32_t max_entries;
+    uint32_t type;
+    std::map<uint32_t, std::unique_ptr<ebpf_map_async_query_context_t>> async_query_contexts;
+    std::condition_variable cleanup_complete_event;
+} ebpf_map_subscription_t;
+
+typedef std::unique_ptr<ebpf_map_subscription_t> ebpf_map_subscription_ptr;
+typedef std::unique_ptr<ebpf_map_async_query_context_t> ebpf_map_async_query_context_ptr;
 
 static ebpf_result_t
-_ebpf_ring_buffer_map_async_query_completion(_Inout_ void* completion_context) NO_EXCEPT_TRY
+_ebpf_map_async_query_completion(_Inout_ void* completion_context) NO_EXCEPT_TRY
 {
     EBPF_LOG_ENTRY();
     ebpf_assert(completion_context);
 
-    ebpf_ring_buffer_subscription_t* subscription =
-        reinterpret_cast<ebpf_ring_buffer_subscription_t*>(completion_context);
+    ebpf_map_async_query_context_t* async_query_context =
+        reinterpret_cast<ebpf_map_async_query_context_t*>(completion_context);
+    ebpf_map_subscription_t* subscription = async_query_context->subscription;
+
+    const uint32_t cpu_id = async_query_context->cpu_id;
 
     size_t consumer = 0;
     size_t producer = 0;
+    size_t lost_count = 0;
 
     ebpf_result_t result = EBPF_SUCCESS;
     // Check the result of the completed async IOCTL call.
-    result = get_async_ioctl_result(subscription->async_ioctl_completion);
+    result = get_async_ioctl_result(async_query_context->async_ioctl_completion);
 
     if (result != EBPF_SUCCESS) {
         if (result != EBPF_CANCELED) {
             // The async IOCTL was not canceled, but completed with a failure status. Mark the subscription object as
             // such, so that it gets freed when the user eventually unsubscribes.
-            std::scoped_lock lock{subscription->lock};
-            subscription->async_ioctl_failed = true;
+            std::scoped_lock lock{async_query_context->lock};
+            async_query_context->async_ioctl_failed = true;
             EBPF_RETURN_RESULT(result);
-        } else {
-            // User has canceled subscription. Invoke user specified callback for the final time with NULL record. This
-            // will let the user app clean up its state.
-            TraceLoggingWrite(
-                ebpf_tracelog_provider,
-                EBPF_TRACELOG_EVENT_GENERIC_MESSAGE,
-                TraceLoggingLevel(WINEVENT_LEVEL_INFO),
-                TraceLoggingKeyword(EBPF_TRACELOG_KEYWORD_API),
-                TraceLoggingString(__FUNCTION__, "ring_buffer map async query completion invoked with EBPF_CANCELED."));
-
-            subscription->sample_callback(subscription->sample_callback_context, nullptr, 0);
         }
```

Comment:
For diagnostics, it will be helpful to tracelog the completion returning a CANCELED result. I also think, the async_query_context should be erased from the list of contexts only if both the following conditions satisfy: 
- callback returned with CANCELED and 
- subscription has been unsubscribed.